### PR TITLE
Tune bitmap scan cost model by updating to nonlinear cost per page

### PR DESCRIPTION
**Pull request by hsyuan**
We had some customers reporting that planner generates plan using seqscan
instead of bitmapscan, and the execution time of seqscan is 5x slower than
using bitmapscan, but the cost of bitmapscan is 2x the cost of seqscan. 
The statistics were updated and quite accurate.

Bitmap table scan uses a formula to interpolate between `random_page_cost` and
`seq_page_cost` to determine the cost per page. But it turns out that the default
value of `random_page_cost` is 100x of the value of `seq_page_cost`. With the
original cost formula, `random_page_cost` predominates in the final cost result,
even the formula is declared to be non-linear, but it is still more like linear,
which can't reflect the real cost per page when a majority of pages are fetched.

Therefore, the cost formula is updated to real non-linear function to reflect
both `random_page_cost` and `seq_page_cost` for different percentage of pages
fetched.

For example, for the default value `random_page_cost = 100`, `seq_page_cost = 1`,
if 80% pages are fetched, the cost per page in old formula is `11.45`, which is
10x more than seqscan, because the cost is dominated by random_page_cost in the
formula. With the new formula, the cost per page is `1.63`, which can reflect the
real cost better, in my opinion.

**Old cost formula:**
```
cost_per_page = random_page_cost - (random_page_cost - seq_page_cost) * sqrt(pages_fetched / T);
```
 
image

**New cost formula:**
```
cost_per_page = seq_page_cost * pow(random_page_cost / seq_page_cost, 1-sqrt(pages_fetched / T));
```

image

[#151934601]

### CODE REVIEW

**Comment by hlinnaka** [comment]
Hmm, yeah, I can see the problem. I find this graph easier to grasp:

save

This plots the total (estimated) cost of a bitmap heap scan, where table size is 10000 pages, and random_page_cost=10 and seq_page_cost=1. X axis is the number of pages fetche. I.e. on the left, no pages are fetched, and on the right end, at 10000, all pages are fetched. The original formula is in black, the new formula in red, and the horizontal line, in blue, shows the cost of a Seq Scan.

In theory, a bitmap heap scan should never be more expensive than a seq scan. In the worst case, the bitmap heap scan needs to scan the whole table, which is essentially a seq scan. So the portion above the blue line doesn't really make sense. In practice, though, a bitmap heap scan probably has a little bit of overhead, compared to a seq scan, just to for bookkeeping if nothing else. So a small overshoot above the blue line probably makes sense. In that light, the new formula looks a lot more sensible. It plateaus just above the cost of a seq scan, which is what I'd expect.

That was with random_page_cost=10, though. If you bump it up further to 100, which is the GPDB default, the graph looks similar in shape, but the new formula plateaus at roughly 2x the cost of the seq scan. Is that appropriate?

I'm a bit worried about the area around 1000-2000. The new formula gives quite different results than the old one there, and I'm not sure which one is more appropriate. Maybe it's OK, but any change to cost estimates, even when they're "right" in principle, is bound to make someone's query choose the wrong plan. Also, the plan changes in the regression suite seem bad. A bitmap index + heap scan of a table with one page and < 20 rows should hardly be cheaper than a seq scan. Maybe this needs some additional fudging for very small tables.

So, this seems reasonable in theory. What about in practice? The underlying assumption here is that as the ratio of pages fetched / total pages goes from 0 to 1, the bitmap heap scan behaves more and more like a seq scan. Is that assumption valid? Or is there some disproportionate cost in "skipping" a few pages in a seq scan, e.g. because it throws off the kernel readahead? If there is, then the original formula actually makes sense.

Is there a difference between heap and AO tables, i.e. is it more expensive to skip "pages" (the block numbers in AO tids don't correspond to physical blocks) in an bitmap AO scan than in a bitmap heap scan?


This isn't really GPDB specific, though. It hurts GPDB particularly badly because of the default random_page_cost is much higher than the Postgres default, but the same problem exists there if you change the defaults. So can you take this to pgsql-hackers, please?

File: src/test/regress/expected/gporca.out [hunk 1]
```diff
@@ -8195,12 +8195,14 @@ set optimizer_enable_constant_expression_evaluation=on;
...
 explain select * from orca.t_date where user_id=9;
-                                          QUERY PLAN                                          
-----------------------------------------------------------------------------------------------
- Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..3105.00 rows=201 width=68)
-   ->  Append  (cost=0.00..3105.00 rows=67 width=68)
-         ->  Seq Scan on t_date_1_prt_part201203 t_date  (cost=0.00..517.50 rows=12 width=68)
-               Filter: user_id = 9::numeric
+                                               QUERY PLAN                                               
+--------------------------------------------------------------------------------------------------------
+ Gather Motion 1:1  (slice1; segments: 1)  (cost=100.61..2931.39 rows=201 width=68)
```
[the remark this answers is not in the document]
Comment (hsyuan) [reply src/test/regress/expected/gporca.out]: The table is not analyzed. I add the analyze table command to update the stats, now there are no plan changes for these queries.

**Comment by hsyuan** [comment]
>That was with random_page_cost=10, though. If you bump it up further to 100, which is the GPDB default, the graph looks similar in shape, but the new formula plateaus at roughly 2x the cost of the seq scan. Is that appropriate?

I think so, because the random cost is so high.

>Also, the plan changes in the regression suite seem bad. A bitmap index + heap scan of a table with one page and < 20 rows should hardly be cheaper than a seq scan. Maybe this needs some additional fudging for very small tables.

I added analyze to update stats in the test, now there are no plan changes.

>Is there a difference between heap and AO tables, i.e. is it more expensive to skip "pages" (the block numbers in AO tids don't correspond to physical blocks) in an bitmap AO scan than in a bitmap heap scan?

@ashwinstar @asimrp Any idea about this?

>This isn't really GPDB specific, though. It hurts GPDB particularly badly because of the default random_page_cost is much higher than the Postgres default, but the same problem exists there if you change the defaults. So can you take this to pgsql-hackers, please?

Sure, I sent it out to pgsql-hackers.

File: src/backend/optimizer/path/costsize.c [hunk 2]
```diff
@@ -716,8 +716,8 @@ cost_bitmap_heap_scan(Path *path, PlannerInfo *root, RelOptInfo *baserel,
 	 * cost per page.
 	 */
 	if (pages_fetched >= 2.0)
-		cost_per_page = random_page_cost -
```
[the remark this answers is not in the document]
Comment (hsyuan) [reply src/backend/optimizer/path/costsize.c]: I have updated the comment. please take a look. thanks

**Comment by hsyuan** [comment]
>In theory, a bitmap heap scan should never be more expensive than a seq scan. In the worst case, the bitmap heap scan needs to scan the whole table, which is essentially a seq scan. So the portion above the blue line doesn't really make sense. 

I think it is quite possible that bitmap scan can be more expensive than seq scan if the fetched pages are large. 

For example, we have 10 pages for a table:
1_2_3_4_5_6_7_8_9_10

If we do seq scan, the estimated total cost for the whole table is 10 (seq_cost = 1). If we bitmap heap scan and the bitmap index selects 5 pages, the total cost may differentiate according to how the selected pages are correlated. If the 5 pages are contiguous, then the total cost is 5. But if the selected pages are 2,4,6,8,10, then every page fetch is random fetch (kind of), the total cost should be near 50, in theory (random_cost = 10). The more percentages the pages are selected, the more possibility some of the selected pages are contiguous.

However, it may be better if we can take `correlation` of pg_stats into consideration when costing bitmap scan.

**Comment by hsyuan** [comment]
Pushed to master, thanks for reviewing!

**Comment by hlinnaka** [comment]
Please add a comment explaining the difference vs upstream.

**Comment by hsyuan** [comment]
Sorry, I forgot to squash these commits.

**Comment by hsyuan** [comment]
I added comment.

**Comment by hlinnaka** [comment]
Oh, never mind, the comment was updated as a separate commit. Please remember to squash before pushing.

Thanks!

**Pull request merged**